### PR TITLE
Optimize Base#relevant_rubocop_rspec_file?

### PR DESCRIPTION
**Pull request by bquorning**
Another take on #965.

The result of `rspec_pattern` is always the same (because it is based on the configuration file), so there is no need to re-calculate it every time it is called.

I tried memozing on the ~Cop~ Base instance, but at least when running the specs, we instantiate ~Cop~ Base way too often. Memoizing on the class level reduces the number of calculations to once per `Base` subclass we have. If we allowed using class variables (`@@rspec_pattern`) we could reduce the number of calculations to one.

This commit reverts 995b4fe5f16cd73feb2625eadd0fad236cfab38e and adds memoization in a class instance variable.

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [-] Added tests.
* [-] Updated documentation.
* [-] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

### CODE REVIEW

**Comment by bquorning** [comment]
And of course this doesn‘t work with specs (spec/rubocop/cop/rspec/cop_spec.rb) that change the configuration.

File: lib/rubocop/cop/rspec/cop.rb [hunk 1]
```diff
@@ -41,32 +33,28 @@ def relevant_file?(file)
...
-        end
-
-        def rspec_pattern_config
-          all_cops_config
-            .fetch('RSpec', DEFAULT_CONFIGURATION)
-            .fetch('Patterns')
+            Config.new
+              .for_all_cops
+              .fetch('RSpec', default_configuration)
+              .fetch('Patterns')
```
Comment (pirj) [inline lib/rubocop/cop/rspec/cop.rb:56]: Isn't the configuration merged with the default configuration, can we assume that `AllCops/RSpec/Patterns` is always present?

File: lib/rubocop/cop/rspec/cop.rb [hunk 2]
```diff
@@ -41,32 +33,28 @@ def relevant_file?(file)
         private
 
         def relevant_rubocop_rspec_file?(file)
-          rspec_pattern =~ file
+          self.class.rspec_pattern =~ file
```
Comment (marcandre) [inline lib/rubocop/cop/rspec/cop.rb:36]: Use `match?`

Comment (pirj) [reply lib/rubocop/cop/rspec/cop.rb:36]: We're introducing `rubocop-performance` to tackle that https://github.com/rubocop-hq/rubocop-rspec/pull/970

File: lib/rubocop/cop/rspec/cop.rb (hunk 1 above)
Comment (bquorning) [reply lib/rubocop/cop/rspec/cop.rb:56]: You are right. We can just use `ConfigLoader.default_configuration.for_all_cops` here.

Comment (bquorning) [reply lib/rubocop/cop/rspec/cop.rb:56]: Still, I’m struggling to figure out how to make this work with the changing configurations in cop_spec.rb

Comment (pirj) [reply lib/rubocop/cop/rspec/cop.rb:56]: Just found this https://github.com/rubocop-hq/rubocop/blob/8842e13621faa82c2688e6d5e7cd891e0aa5873c/lib/rubocop/rspec/shared_contexts.rb#L76
Maybe can be of some help?

Comment (pirj) [reply lib/rubocop/cop/rspec/cop.rb]: I mean it could be part of 2.0, not even post-2.0. We have a bunch of pull requests specifically for 2.0.

**Comment by bquorning** [comment]
Fixed in #1063.

**Pull request closed without merging**